xcdat/bounds.py

```diff
@@ -429,3 +432,227 @@ def _validate_axis_arg(self, axis: CFAxisKey):
+        2, 3, 4, 6, 8, 12, or 24.
+
+    Notes
+    -----
+    This function is intended to reproduce CDAT's setAxisTimeBoundsDaily
+    method [1]_.
+
+    References
+    ----------
+    [1] https://github.com/CDAT/cdutil/blob/master/cdutil/times.py#L1093
+    """
+    # get calendar
+    calendar = time.encoding["calendar"]
+
+    # get cftime class to create new cftime objects
+    cf_obj = get_date_type(calendar)
+
+    # loop over time values and compute bounds
+    time_bnds = []
+    if (frequency > 24) | (np.mod(24, frequency)):
+        raise ValueError(
+            "Incorrect `frequency` argument."
+            " Supported values include 2, 3, "
+            "4, 6, 8, 12, and 24."
+        )
+    for step in time.values:
+        # get year / month
+        y, m, d, h = step.year, step.month, step.day, step.hour
+        for f in range(frequency):
+            if f * (24 // frequency) <= h < (f + 1) * (24 // frequency):
+                l_bnd = cf_obj(y, m, d, f * (24 // frequency))
+                u_bnd = l_bnd + datetime.timedelta(hours=(24 // frequency))
+        # store
+        time_bnds.append([l_bnd, u_bnd])
+
+    # create dataarray
+    time_bnds = xr.DataArray(  # type: ignore
+        name=f"{time.name}_bnds",
+        data=time_bnds,
+        coords=dict(time=time),
+        dims=[*time.dims, "bnds"],
+        attrs={"xcdat_bounds": "True"},
+    )
+
+    return time_bnds
+
+
+def get_time_bounds(time):
+    """Sets the time bounds for a time coordinate axis.
+
+    Parameters
+    ----------
+    time : xr.DataArray
+        The temporal coordinate variable for the axis.
+    frequency : int, optional, default 1
+        Flag to note set the number of timepoints per day.
+
+    Returns
+    -------
+    xr.DataArray
+        The bounds for the time axis.
+
+    Notes
+    -----
+    Function uses `_infer_freq` to determine the temporal resolution
+    and the calls `get_yearly_time_bounds`, `get_monthly_time_bounds`,
+    or `get_daily_time_bounds` as appropriate. See these functions for
+    more details.
+    """
+    freq = _infer_freq(time)
+    if freq == "year":
+        time_bnds = get_yearly_time_bounds(time)
+    if freq == "month":
+        time_bnds = get_monthly_time_bounds(time)
+    if freq == "day":
+        time_bnds = get_daily_time_bounds(time)
+    if freq == "hour":
+        time_bnds = get_daily_time_bounds(time)
+        # get number of time steps per day
+        diff = time.values[1] - time.values[0]
+        hrs = diff.seconds / 3600
+        daily_freq = int(24 / hrs)
+        # get sub-daily time bounds
+        time_bnds = get_daily_time_bounds(time, frequency=daily_freq)
+
+    # Create the bounds data array
+    time_bnds = xr.DataArray(
+        name=f"{time.name}_bnds",
+        data=time_bnds,
+        coords={time.name: time},
+        dims=[*time.dims, "bnds"],
+        attrs={"xcdat_bounds": "True"},
+    )
+
+    return time_bnds
```

Comment:
@pochedls I had a few questions about these new time bounds APIs:

1. Does you think it makes sense to define these functions as `BoundsAccessor` methods? 
2. Are they intended to be public APIs?
3. Does it make sense to prepend the API names with `create_` instead of `get_`?
   - `add_bounds()` calls `_create_bounds()`, which calls `get_time_bounds()`. Maybe renaming `get_time_bounds()` to `_create_time_bounds()` is more appropriate?
   - There might be a bit of confusion between the usage for the `get_bounds()` method and the `get_time_bounds()` function because they both operate on the time axis.

Reply:
> 1. Does you think it makes sense to define these functions as BoundsAccessor methods?

These are currently not `BoundsAccessor` methods, right?

I think I initially made them accessor methods and then felt I was making life complicated. Accessor methods were either a) unnecessary (if you pass in the time coordinate, you don't need the `dataset` object) or b) a lot of extra work (because then the method has to figure out what time coordinate axis you care about in each of these functions). I guess you could also pass in the `.name` attribute of the time coordinate you care about (and then use the dataset to fetch the time coordinate axis), but then why not just pass in the time axis? 

> 2. Are they intended to be public APIs? 

Yes – I think they should be accessible. They seem like they could be useful functions for users to be able to call. 

> 3. Does it make sense to prepend the API names with `create_` instead of `get_`? 

I think you are right.

Reply:
> I think I initially made them accessor methods and then felt I was making life complicated. 

From what I understand, there are two scenarios here:

1.  We want to add time bounds directly to the Dataset - use `ds.temporal.add_bounds(axis="T")`
    - Loops over _all_ time coordinates and creates bounds for each (calls `create_time_bounds()`)
    - Returns an `xr.Dataset`
    - Example usage: `ds = ds.bounds.add_bounds(axis="T")`
2.  We want to create time bounds for a _specific_ time coordinate variable for more granular operations - use `create_time_bounds()`
    - User specifies the exact time coordinate variable to create bounds.
    - Returns an `xr.DataArray`
    - Example usage: `time_bounds = create_time_bounds(ds.time)`

---

If we want to support scenario 2, I can see how the implementation/usage of `create_time_bounds()` is cumbersome if it were a `BoundsAccessor` method. `BoundsAccessor` methods are intended to return an `xr.Dataset`, not an `xr.DataArray`.

If this all sounds correct, it makes sense to keep your implementation of `create_time_bounds()` as a **function**.

Reply:
I think scenario 1 would be ideal, but I think it would also be useful to allow users to call the methods individually. CDAT didn't automatically create time bounds (as far as I know) and I worry there might be situations where we cannot automatically infer the frequency and add the time bounds. In those cases, enabling users to choose between create_daily/monthly/yearly_bounds and create_time_bounds would be helpful. Doesn't the code, as written, allow you to do both (call individual functions or infer the bounds through `add_bounds()`)? Is there a better way to enable both? 

Semi-relevant: I had initially modified [`create_bounds` to accept a dataarray](https://github.com/xCDAT/xcdat/pull/418/commits/ffb2688d7f33ed1362a8b335e00e2090b548d6b7#diff-6a74d16440d82ae43ed458dc26fafd0427c8568b34915e698fce57b2c9a8a7e1R232) if the user wanted to supply their own bounds (e.g., from `create_monthly_time_bounds`). I undid this since I expected bounds would generally be added through `add_bounds`.

Reply:
Thanks for clarifying Steve. It sounds like we're on the same page in regards to the usefulness of supporting both scenarios. 

I'll try to see if there is an intuitive way of implementing and using `create_time_bounds()` as a `BoundsAccessor` method. If I can't find one, I'm happy to keep your function-based implementation.

Reply:
After some thought, I wasn't able to find an _intuitive_ accessor method implementation of `create_time_bounds()`. I am leaning towards keeping your implementation.

### 1. Function approach (current)

Example:
```python

def create_time_bounds(time: xr.DataArray) -> xr.DataArray
	pass

def create_yearly_time_bounds(self, xr.DataArray) -> xr.DataArray
	pass

def create_monthly_time_bounds(self, xr.DataArray) -> xr.DataArray
	pass

def create_daily_time_bounds(self, xr.DataArray) -> xr.DataArray
	pass

ds["time_bnds"] = create_time_bounds(ds.time)
ds["time_bnds"] = create_yearly_time_bounds(ds.time)
ds["time_bnds"] = create_monthly_time_bounds(ds.time)
ds["time_bnds"] = create_daily_time_bounds(ds.time)
```

Pros:
  - Simpler API naming convention (no need to reference `ds.bounds`)
  - User references the coordinate variable directly rather than xCDAT checking the dataset for the coordinate variable  -- earlier validation in the pipeline
  - `ds.bounds.add_bounds()` will handle _most_ cases for adding bounds -- these functions are for edge cases or users who want finer grain control

Cons:
 - Functionality is related to the `BoundsAccessor` class so grouping kinda makes sense

### 2. Accessor approach

Example:
```python
def create_time_bounds(self, time: str) -> xr.DataArray
	pass

def create_yearly_time_bounds(self, time:str) -> xr.DataArray
	pass

def create_monthly_time_bounds(self, time:str) -> xr.DataArray
	pass

def create_daily_time_bounds(self, time:str) -> xr.DataArray
	pass

ds["time_bnds"] = ds.bounds.create_time_bounds("time")
ds["time_bnds"] = ds.bounds.create_yearly_time_bounds("time")
ds["time_bnds"] = ds.bounds.create_monthly_time_bounds("time")
ds["time_bnds"]  = ds.bounds.create_daily_time_bounds("time")
```
Pros:
  - Encapsulates related bounds methods together
  - References the dataset and attempts to grab `ds.time` coordinate variable

Cons:
  - Verbose method naming convention makes it tedious to call when it's under the `.bounds` namespace
  - Returns a DataArray -- user needs to set bounds as data variable in the Dataset regardless if it were method or a function (e.g., `ds["time_bnds"] = ...`) 
  - We don't reference any class attribute in the methods except for `self._dataset` to get the `ds.time` coordinate variable